File: Structural/Decorator/cpp/test_decorator.cpp
#include "decorator.hpp"
#include <gtest/gtest.h>

/**
 * @brief Test the base ConcreteComponent.
 */
TEST(DecoratorPatternTest, ConcreteComponent) {
	auto component = std::make_shared<ConcreteComponent>();
	EXPECT_EQ(component->getDescription(), "Concrete Component");
	EXPECT_DOUBLE_EQ(component->cost(), 10.0);
}

/**
 * @brief Test the functionality of ConcreteDecoratorA.
 */
TEST(DecoratorPatternTest, ConcreteDecoratorA) {
	auto component = std::make_shared<ConcreteComponent>();
	auto decorated = std::make_shared<ConcreteDecoratorA>(component);

	EXPECT_EQ(decorated->getDescription(), "Concrete Component + Feature A");
	EXPECT_DOUBLE_EQ(decorated->cost(), 15.0);
}

/**
 * @brief Test the functionality of ConcreteDecoratorB.
 */
TEST(DecoratorPatternTest, ConcreteDecoratorB) {
	auto component = std::make_shared<ConcreteComponent>();
	auto decorated = std::make_shared<ConcreteDecoratorB>(component);

	EXPECT_EQ(decorated->getDescription(), "Concrete Component + Feature B");
	EXPECT_DOUBLE_EQ(decorated->cost(), 17.5);
}

/**
 * @brief Test chaining multiple decorators.
 */
TEST(DecoratorPatternTest, MultipleDecorators) {
	auto component = std::make_shared<ConcreteComponent>();
	auto decoratorA = std::make_shared<ConcreteDecoratorA>(component);
	auto decoratorB = std::make_shared<ConcreteDecoratorB>(decoratorA);

	EXPECT_EQ(decoratorB->getDescription(), "Concrete Component + Feature A + Feature B");
	EXPECT_DOUBLE_EQ(decoratorB->cost(), 22.5);
}

int main(int argc, char **argv) {
	::testing::InitGoogleTest(&argc, argv);
	return RUN_ALL_TESTS();
}

File: Structural/Decorator/python/test_decorator.py
import unittest
from decorator import Component, ConcreteComponent, ConcreteDecoratorA, ConcreteDecoratorB


class TestDecoratorPattern(unittest.TestCase):
	"""
	Unit tests for the Decorator Pattern implementation.
	"""

	def test_concrete_component(self):
		"""
		Test that the ConcreteComponent provides the correct description and cost.
		"""
		component = ConcreteComponent()
		self.assertEqual(component.get_description(), "Concrete Component")
		self.assertEqual(component.cost(), 10.0)

	def test_concrete_decorator_a(self):
		"""
		Test that the ConcreteDecoratorA correctly adds feature A to the decorated component.
		"""
		component = ConcreteComponent()
		decorator_a = ConcreteDecoratorA(component)

		self.assertEqual(decorator_a.get_description(), "Concrete Component + Feature A")
		self.assertEqual(decorator_a.cost(), 15.0)

	def test_concrete_decorator_b(self):
		"""
		Test that the ConcreteDecoratorB correctly adds feature B to the decorated component.
		"""
		component = ConcreteComponent()
		decorator_b = ConcreteDecoratorB(component)

		self.assertEqual(decorator_b.get_description(), "Concrete Component + Feature B")
		self.assertEqual(decorator_b.cost(), 17.5)

	def test_multiple_decorators(self):
		"""
		Test that multiple decorators can be applied to a component.
		"""
		component = ConcreteComponent()
		decorator_a = ConcreteDecoratorA(component)
		decorator_b = ConcreteDecoratorB(decorator_a)

		self.assertEqual(decorator_b.get_description(), "Concrete Component + Feature A + Feature B")
		self.assertEqual(decorator_b.cost(), 22.5)


if __name__ == "__main__":
	unittest.main()
